svc/FetchService.ts

```diff
@@ -79,24 +107,29 @@ export class FetchService extends HoistService {
      * Send a request via the underlying fetch API.
      * @returns Promise which resolves to a Fetch Response.
      */
-    async fetch(opts: FetchOptions): Promise<FetchResponse> {
-        opts = await this.withDefaultsAsync(opts);
-        return this.managedFetchAsync(opts);
+    fetch(opts: FetchOptions): Promise<FetchResponse> {
+        opts = this.withResolvedCorrelationId(opts);
+        const ret = this.withDefaultHeadersAsync(opts).then(opts => this.managedFetchAsync(opts));
+        ret.correlationId = opts.correlationId as string;
+        return ret;
     }
 
     /**
      * Send an HTTP request and decode the response as JSON.
      * @returns the decoded JSON object, or null if the response has status in {@link NO_JSON_RESPONSES}.
      */
-    async fetchJson(opts: FetchOptions): Promise<any> {
-        opts = await this.withDefaultsAsync(opts, {Accept: 'application/json'});
-        return this.managedFetchAsync(opts, async r => {
-            if (this.NO_JSON_RESPONSES.includes(r.status)) return null;
-
-            return r.json().catchWhen('SyntaxError', e => {
-                throw Exception.fetchJsonParseError(opts, e);
-            });
-        });
+    fetchJson(opts: FetchOptions): Promise<any> {
+        opts = this.withResolvedCorrelationId(opts);
+        const ret = this.withDefaultHeadersAsync(opts, {Accept: 'application/json'}).then(opts =>
+            this.managedFetchAsync(opts, async r => {
+                if (this.NO_JSON_RESPONSES.includes(r.status)) return null;
+                return r.json().catchWhen('SyntaxError', e => {
+                    throw Exception.fetchJsonParseError(opts, e);
+                });
+            })
+        );
+        ret.correlationId = opts.correlationId as string;
```

Comment:
Do we have any sense that APIs might be returning correlation IDs in a response header and that we should provide a way to specify that header and have this code alternately check it?

I looked at one API in a client project that I know generates IDs without us specifying them (as we see them in error messages), and I do _not_ see those IDs returned in headers, so perhaps this is not a common pattern.  And in general if we are setting a corrId on the outbound request ourself, the entire idea is that the same ID should come back to us if an API were to be including them.  

So I guess I am imagining a case where we are not spec'ing IDs but the API is generating them _and_ returning them. Probably not anything we need to worry about, just wanted to flag as a possible consideration in case it ever comes up.

Reply:
That's an interesting case you also mentioned in your original ticket.  I'm not sure what would be best, so just listing some scenarios to consider when discussing this:

Scenario A:
- Client makes a call and does not provide a Correlation ID.
- External API assigns a Correlation ID to the request, but client times out before the response is sent.

In the above case, the Correlation ID isn't very effective since we have no record of it on the client.  Seems like ideally we want the originator of a request to be assigning the correlation ID. As I write this, I'm realizing we should probably think about how correlation ID's might factor into our websocket service (if at all) where the roles are reversed.

Scenario B:
- Client makes a call and provides a Correlation ID
- External API responds with some other Correlation ID.

In the above case, do we warn? Track / log both IDs?

Reply:
Agreed that in general the idea is that we originate correlation IDs (at least as far as fetchService is concerned). 

I don't think that this is likely, but it seems at least possible that we would have some project where, for whatever reason, the API wanted to take control of this and requested that we didn't send outbound IDs but did return them in a header. 

The scenario B above where the API misbehaves and returns an alternate ID seems like it's just an error - i.e. we are sending the outbound request in some other header than what the API expects.

More than anything, it felt like there was something to be said for reading and respecting any ID coming back on a request header.  BUT I then went and looked at two easy-to-check APIs  - the MSAL call to get tokens and a client-app specific API that I know uses recent .net and returns corrIDs with errors.  Neither of those had a correlation ID header on the response or returned one in the body, so maybe that's just "not a thing," or at least I'm no longer motivated.  We can leave as-is and revisit if we have a real use case.